drivers/flash/disk_flash.c

```diff
@@ -0,0 +1,344 @@
+{
+	const struct disk_flash_config *cfg = dev->config;
+
+	return &cfg->flash_parameters;
+}
+
+static const struct flash_driver_api disk_flash_api = {
+	.read = disk_flash_read,
+	.write = disk_flash_write,
+	.erase = disk_flash_erase,
+	.get_parameters = disk_flash_get_parameters,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = disk_flash_pages_layout,
+#endif
+#if defined(CONFIG_FLASH_JESD216_API)
+	.sfdp_read = NULL,
+	.read_jedec_id = NULL,
+#endif
+#if defined(CONFIG_FLASH_EX_OP_ENABLED)
+	.ex_op = NULL,
+#endif
+};
+
+static int disk_flash_init(const struct device *dev)
+{
+	int res = 0;
+	struct disk_flash_data *data = dev->data;
+	const struct disk_flash_config *cfg = dev->config;
+	size_t disk_size = 0;
+
+	k_sem_init(&data->sem, 1, 1);
+
+	res = disk_access_init(cfg->disk_name);
+	if (res != 0) {
+		LOG_ERR("init disk failed: %d", res);
+		goto init_done;
+	}
+
+	res = disk_access_ioctl(cfg->disk_name, DISK_IOCTL_GET_SECTOR_COUNT,
+				&data->disk_total_sector_cnt);
+	if (res != 0) {
+		LOG_ERR("read total sector cnt failed: %d", res);
+		goto init_done;
+	}
+
+
+	res = disk_access_ioctl(cfg->disk_name, DISK_IOCTL_GET_SECTOR_SIZE,
+				&data->disk_sector_size);
+	if (res != 0) {
+		LOG_ERR("read sector size failed: %d", res);
+		goto init_done;
+	}
+
+	if (data->disk_sector_size > data->sector_buf_size) {
+		LOG_ERR("sector size %u of disk to big for buffer %u", data->disk_sector_size,
+			(uint32_t)data->sector_buf_size);
+		res = -EINVAL;
+		goto init_done;
+	}
+
+	if (cfg->disk_offset % data->disk_sector_size != 0) {
+		LOG_WRN("offset is not aligned with disk sectors (disk sector size: %u)",
+			data->disk_sector_size);
+	}
+
+	disk_size = data->disk_sector_size * data->disk_total_sector_cnt;
+	if ((disk_size - cfg->disk_offset) < cfg->flash_size) {
+		LOG_ERR("Underlying disk to small to support flash: %lu < %d",
+			(unsigned long)(disk_size - cfg->disk_offset),
+			cfg->flash_size);
+		res = -EINVAL;
+		goto init_done;
+	}
+
+init_done:
+	return res;
+}
+
+#define DISK_FLASH_DEFINE(index)								\
+	BUILD_ASSERT(DT_INST_PROP(index, size) > 0, "flash_size needs to be > 0");		\
+	BUILD_ASSERT(DT_INST_PROP(index, disk_offset) >= 0, "disk_offset needs to be >= 0");	\
+	uint8_t sector_buf_##index[DT_INST_PROP(index, disk_sector_size)];			\
+	static struct disk_flash_data disk_flash_data_##index = {				\
+		.sector_buf = sector_buf_##index,						\
+		.sector_buf_size = sizeof(sector_buf_##index),					\
+	};											\
+	static const struct disk_flash_config disk_flash_config_##index = {			\
+		.flash_size = DT_INST_PROP(index, size),					\
+		.flash_parameters =								\
+		{										\
+			.write_block_size = 1,							\
+			.erase_value = 0xFF							\
+		},										\
+		IF_ENABLED(									\
+			CONFIG_FLASH_PAGE_LAYOUT,						\
+			(.fpl =									\
+				{								\
+					.pages_count = DT_INST_PROP(index, size),		\
+					.pages_size = 1						\
```

Comment:
Erase at least by sector size.

Reply:
Is there a reason to make this driver dependent on the disk sector size? This driver currently supports arbitrary write/read/erase sizes independent from the underlying disk.